chain/ethereum/ethereum.go

```diff
@@ -266,17 +268,44 @@ func HistoricalEvents(ctx context.Context, client *ethclient.Client, addresses [
 		//	Store failed queries for retry
 		failedQueries, err := ChunkedFilterLogs(ctx, client, addresses, fromBlock, toBlock, logs, nil)
 		if err != nil {
-			log.Error().Uint64("from", fromBlock).Uint64("to", toBlock).Msg("some intervals failed during backfill..")
+			log.Warn().Err(err).Uint64("from", fromBlock).Uint64("to", toBlock).Msg("some intervals failed during backfill, retying...")
 		}
 
-		for _, q := range failedQueries {
-			//	Retry failed queries one more time
+		// Count the number of consecutive failures. Reset the counter after every success.
+		// If fail 2 times consecutively, start doing exponential backoff before retry again.
+		failCount := 0
+
+		for len(failedQueries) > 0 {
+			// Get a failed query from the beginning of the queue
+			q := failedQueries[0]
+			failedQueries = failedQueries[1:]
+
 			fq, err := ChunkedFilterLogs(ctx, client, q.Addresses, q.FromBlock.Uint64(), q.ToBlock.Uint64(), logs, nil)
 			if err != nil {
-				log.Error().Uint64("from", q.FromBlock.Uint64()).Uint64("to", q.ToBlock.Uint64()).Msg("some intervals failed during backfill retry..")
+				log.Warn().Err(err).Uint64("from", q.FromBlock.Uint64()).Uint64("to", q.ToBlock.Uint64()).Msg("some intervals failed during backfill, retrying...")
 			}
-			for _, q2 := range fq {
-				log.Error().Uint64("from", q2.FromBlock.Uint64()).Uint64("to", q2.ToBlock.Uint64()).Msg("aborting failed backfill interval.")
+
+			if len(fq) > 0 {
+				// Put failed queries to the end of queue to retry later
+				failedQueries = append(failedQueries, fq...)
+				failCount++
+			} else {
+				failCount = 0
+			}
+
+			if failCount >= 2 {
+				// Do exponential backoff with 10% jitter
+				backoff := float64(int(1) << (failCount - 2))
```

Comment:
Can you put some comments to explain this logic? It's not easy to understand.

Reply:
So basically I am calculating the backoff duration after each failure. For the first 2 failures, there will be no backoff. After that, each failure will increase the backoff duration by double starting from 1 second. I also add 10% jitter for the backoff duration to avoid calling PRCs at predictable times.

The reason I dont backoff for the first 2 failures is because after one failure, I dont want to wait before executing the next query.

I am not sure that I should put these explainations in the comment or not since they are pretty long and unnecessary.

Reply:
Can you also make this `2` a constant?

Reply:
Update the comment to explain the code better. 
2 will not be configurable so it is not necessary to define a constant.